src/bsoncxx/document/value.hpp

```diff
@@ -76,6 +81,25 @@ class BSONCXX_API value {
     value(value&&) noexcept = default;
     value& operator=(value&&) noexcept = default;
 
+    ///
+    /// Constructor used for serialization of user objects. This uses argument-dependent lookup
+    /// to find the function declaration `void to_bson(T& t, bsoncxx::document::value doc)`.
+    ///
+    /// @param user_object
+    ///   A custom user object to serialize. This object must be default-constructible for
+    ///   deserialization.
+    ///
+    template <typename T,
+              typename std::enable_if<!std::is_same<T, typename array::view>::value, int>::type = 0>
+    explicit value(const T& user_object) : value({}) {
+        to_bson(user_object, *this);
+    }
+    template <typename T>
+    value& operator=(const T& user_object) {
```

Comment:
nit: You could remove the "user" in "user_object" here and above. It doesn't matter who owns or create the object. We usually just call this parameter _t_ or _v_ or even _value_ depending on the context.